jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/StepsRunner.java

```diff
@@ -306,36 +311,62 @@ private void obtainBaseImagesLayers(
                       "scheduling obtaining base images layers",
                       results.baseImagesAndRegistryClient.get().images.size())) {
 
+                Map<DescriptorDigest, Future<PreparedLayer>> preparedLayersCache = new HashMap<>();
                 Map<Image, List<Future<PreparedLayer>>> baseImagesAndLayers = new HashMap<>();
                 for (Image baseImage : results.baseImagesAndRegistryClient.get().images) {
                   List<Future<PreparedLayer>> layers =
                       obtainBaseImageLayers(
-                          baseImage, layersRequiredLocally, progressDispatcher.newChildProducer());
+                          baseImage,
+                          layersRequiredLocally,
+                          preparedLayersCache,
+                          progressDispatcher.newChildProducer());
                   baseImagesAndLayers.put(baseImage, layers);
                 }
                 return baseImagesAndLayers;
               }
             });
   }
 
-  private List<Future<PreparedLayer>> obtainBaseImageLayers(
+  // This method updates the given "preparedLayersCache" and should not be called concurrently.
+  @VisibleForTesting
+  List<Future<PreparedLayer>> obtainBaseImageLayers(
       Image baseImage,
       boolean layersRequiredLocally,
+      Map<DescriptorDigest, Future<PreparedLayer>> preparedLayersCache,
       ProgressEventDispatcher.Factory progressDispatcherFactory)
       throws InterruptedException, ExecutionException {
-    return scheduleCallables(
-        layersRequiredLocally
-            ? ObtainBaseImageLayerStep.makeListForForcedDownload(
-                buildContext,
-                progressDispatcherFactory,
-                baseImage,
-                results.baseImagesAndRegistryClient.get().registryClient)
-            : ObtainBaseImageLayerStep.makeListForSelectiveDownload(
-                buildContext,
-                progressDispatcherFactory,
-                baseImage,
-                results.baseImagesAndRegistryClient.get().registryClient,
-                results.targetRegistryClient.get()));
+    List<Future<PreparedLayer>> preparedLayers = new ArrayList<>();
+
+    try (ProgressEventDispatcher progressDispatcher =
+        progressDispatcherFactory.create(
+            "launching base image layer pullers", baseImage.getLayers().size())) {
+      for (Layer layer : baseImage.getLayers()) {
+        DescriptorDigest digest = layer.getBlobDescriptor().getDigest();
+        Future<PreparedLayer> preparedLayer = preparedLayersCache.get(digest);
+
+        if (preparedLayer != null) {
+          progressDispatcher.dispatchProgress(1);
```

Comment:
Just a thought: we currently check to see if the layer already exists before pushing a new one, but what would we do if we needed to update a layer?/ Can base image layers be updated?

Reply:
Not sure I understand your point. To clarify a few things,
- This section of code is to *download* a base image layer.
- In a later stage, base image layers need to be pushed to a target registry (but only those layers missing in the registry).
- A layer is just a block of bytes (blob) identified and addressed by a SHA digest. And every blob is immutable and content-addressable. There's nothing like "updating" a blob or an existing image. If a content of a blob or an "image" is different, it is a new blob or a new image with different addresses.

Reply:
I see, thanks for clarifying! This really clears up some of the key concepts I was missing:)